### ISSUE TITLE
Getting of the Aphrodite class

### ISSUE THREAD

**Issue by rpelisse** [jboss-set/aphrodite#162]
@wolfc @TomasHofman @baranowb @soul2zimate let's keep our discussion on how to get ride off this large and clunky Aphrodite...

**Comment by TomasHofman** [comment]
The original question was 
> Should we split Aphrodite class into separate "repository" or "service" classes?

Perhaps the `Aphrodite` class could stay to hold some common resources such us the `ExecutorService` a to provide access to service classes that would provide entity-related behaviour.

What may also affect this refactoring is how we want to implement the access to related entities, ie `issue.getPatches()`. I think it was suggested during the tech talk to make entities "smart" and give them ability to load related entities by themselves, like in:

```
class Issue {
    public Stream<Patch> getPatches() throws NameNotFoundException {
        return Container.instance().lookup(PatchHome.class.getSimpleName(), (PatchHome.class)).findPatchesByIssue(this);
    }
}
```

which effectively gives the an entity access to a repository / domain service. It looks like violating the single responsibility principle, because now an entity (which should be responsible only for its' own state) knows something about how to load other entities. It's also form of lazy loading.

What about:

1. Keeping related entities loading to domain services, like `aphrodite.getPullRequestService().getPullRequestsForIssue(issue)`.
2. Implement "loaders" that would load whole entity tree and prefill the fields upon asking, like `aphrodite.getIssueLoader().includePullRequests().includeViolations().getIssues(listOfUrls)`.
3. If we want to have lazy loading, do it more like:
```
public class Issue {
    private Supplier<List<PullRequest>> pullRequestSupplier;
    public void setPullRequestSupplier(Supplier<List<PullRequest>> pullRequestSupplier) {
        this.pullRequestSupplier = pullRequestSupplier;
    }
    public List<PullRequest> getPullRequests() {
        return pullRequestSupplier.get();
    }
}
```
where `pullRequestSupplier` is supplied by a domain service and encapsulates the logic behind loading subentities.

The first two options could maybe allow some optimizations as subentities could be loaded in bulks.

**Comment by wolfc** [comment]
> It looks like violating the single responsibility principle, because now an entity (which should be responsible only for its' own state) knows something about how to load other entities. It's also form of lazy loading.
It delegates the responsibility of loading a service to `Container` and handling `Patch` to `PatchHome`.
It should actually use `@Inject` to get to those services. :)

**Comment by wolfc** [comment]
I don't see a need to hang onto a common `ExecutorService`. Anything executed in parallel should make use of the standard JDK facilities.

**Comment by TomasHofman** [comment]
Yes, but the point was that entity shouldn't(?) need to know anything about PatchHome nor Container. Also the container module (functional thing) then needn't be a dependency of the domain module.

No matter if it's obtained via @Inject or Container, it seems unusual to me that an entity is accessing a repository.

**Comment by rpelisse** [comment]
Here my thought about something somewhat related. Right now, BugClerk just needs to use Aphrodite to retrieve issues. With Janitor being added as a service, BugClerk will just need to get an instance IssueTrackerService and one of Janitor. So, BugClerk does not need to use the all Aphrodite class, and could get away with doing something along the lines of :
``` 
ServiceLoader<IssueTrackerService> trackerLoader = ServiceLoader.load(IssueTrackerService.class);
ServiceLoader<JanitorService> janitorLoader = ServiceLoader.load(JanitorService.class);
        BugClerk bugclerk = new BugClerk(janitorLoader.iterator().next(), trackerLoader.iterator().next());
```
However, the instance obtained this way are not configured (no tracker url for the IssueTrackerService for instance).  

Maybe I am wrong, but won't be better (and also necessary if we want to move away from the large Aphrodite class approach) to have each service offering an empty init() method configuring itself using the aphrodite-config.json content ?

We will just need to move the following code into an utility classes:

```
private Aphrodite() throws AphroditeException {
        String propFileLocation = System.getProperty(FILE_PROPERTY);
        if (propFileLocation == null)
            throw new IllegalArgumentException("Property '" + FILE_PROPERTY + "' must be set");

        try (JsonReader jr = Json.createReader(new FileInputStream(propFileLocation))) {
            init(AphroditeConfig.fromJson(jr.readObject()));
        } catch (IOException e) {
            Utils.logException(LOG, "Unable to load file: " + propFileLocation, e);
            throw new AphroditeException(e);
        }
    }
```
This way, we could simply retrieve the services using the JDK API, and only use the Aphrodite class when we do need "everything". 

What do you think guys ?

**Comment by wolfc** [comment]
Consider `JiraIssueTracker.init(AphroditeConfig)` deprecated, rather look at `JiraIssueTracker.init(IssueTrackerConfig)`.
I don't like the usage of `ServiceLoader`. That should either be `Container.lookup` or `@Inject`.
CDI should allow for a "stateful" injection. `JiraIssueTracker` should never come up non-functional, so probably the whole story around `init` is moot.

**Comment by rpelisse** [comment]
> Consider JiraIssueTracker.init(AphroditeConfig) deprecated, rather look at JiraIssueTracker.init(IssueTrackerConfig).

I agree with that, what I'm suggesting is adding a init() that will use the code mentioned above to allow the service to initialise themself.

>I don't like the usage of ServiceLoader. 

The same situation comes with the Container.lookup. Currently, the service registered are not configure. So (unless I'm wrong of course), you execute lookup() and you get an instance of, let's say, IssueTrackerService, but it is not configured. You need to pass it IssueTrackerConfig, which means parsing the AphroditeConfig and so on... 

Right now, this job is done inside the Aphrodite class, but (afaiu) if we stop using the Aphrodite class, we will need to all this wiring instead. But maybe I've missed something ?

**Comment by TomasHofman** [comment]
> BugClerk just needs to use Aphrodite to retrieve issues. With Janitor being added as a service, BugClerk will just need to get an instance IssueTrackerService and one of Janitor.

Do you really need just the IssueTrackerService? What if you need to fetch a pull request to perform some checks on it? In the last check that I tried to implement I needed that.

**Comment by rpelisse** [comment]
Yes, BugClerk needs more than the IssueTracker, but my point is about how to do the wiring when using less than all the service in Aphrodite.

Either we go with an injection approach, where we inject part of Aphrodite we need (and thus my question on how to have the service "ready for use" applies) or we don't, and stay with the large Aphrodite class acting as micro-kernel and taking of building all the dependencies.

My understanding is that Carlo wishes to move away from the large Aphrodite class to have a container approach (even maybe use CDI and @Inject). But, right now, if I use the lookup() method, I end up having a not initialized Service, and I don't want to copy the parsing code over to Bugclerk. Maybe,down the road, we will have that fixed by CDI or more advanced container, but right now I don't know how to address it betterly...

**Comment by TomasHofman** [comment]
Should we perhaps create a serious class diagram first, so that we have something tangible to discuss?

What about this as a starting point: https://drive.google.com/open?id=0B-Ee_N1p3arSWjNMdTNkWDJKVVE

If Aphrodite is split into several classes, like IssueManager and PRManager, then some service types may be needed in both managers (for instance IssueManager may need access to both, TrackerServices and RepositoryServices, so that it can preload or inject PRs in issues that are being loaded). So these manager classes are not suitable place to initialize tracker/repository services. Aphrodite class could be the place where initialization is being done and hold references to managers.

Or we could add another layer, like TrackerAggregateService and PRAggregateService, initialize all trackers in former, all repositories in latter, and the aggregate services would be injected into managers that would provide higher level functionality like preloading/injecting of related entities.

BTW I created the above image in Gliffy, and it's by far the most handy tool I found for UML modelling, only one needs to register via Google account or some other account. The diagram should be accessible for modification via this link: https://go.gliffy.com/go/share/s5p7gr9vsvl82haymhfd

**Comment by TomasHofman** [comment]
Version with aggregate services: https://drive.google.com/open?id=0B-Ee_N1p3arSSVhMY0FaeEZKOFE

**Comment by TomasHofman** [comment]
Do anyone agree that split of Aphrodite into "managers" is still needed?

If I were a bugclerk, I would still like to have some way of loading issues / PRs without having to work with specific issue trackers / repository services.

Also these higher level things may be better fit for injecting into entities. E.g. when calling issue.getPullRequests() there may be PRs in various repositories, so entity shouldn't need to worry about specific repository services either.

**Comment by rpelisse** [comment]
IMHO, I don't think have more granularity will hurt us. Worst case
scenario, we just make the Aphrodite class less large and keep using it and
best case scenario, having those separated managers will really help us
pick and choose which part of Aphrodite we need to have.
